src/components/Taquin.astro

```diff
@@ -1,19 +1,266 @@
+        game.moves = 0;
+
+        has_won = false;
+        in_game = false;
+        show_grid = true;
+        can.classList.remove("anim");
+        document.getElementById("floating").classList.remove("anim");
+
+        for (var i = 0; i < shuffle_num; i++) {
+            shuffle_once();
+        }
+        in_game = true;
+
+        draw();
+    }
+
+    function shuffle_once() {
+        var x = Math.floor(Math.random() * game.size);
+        var y = Math.floor(Math.random() * game.size);
+        click({
+            clientX: (x * can.width) / game.size + can.offsetLeft,
+            clientY: (y * can.height) / game.size + can.offsetTop,
+        });
+    }
+
+    function draw() {
+        ctx.fillStyle = hole_color;
+        ctx.fillRect(0, 0, can.width, can.height);
+
+        var w = can.width / game.size;
+        var h = can.height / game.size;
+
+        for (var i = 0; i < game.size; i++) {
+            for (var j = 0; j < game.size; j++) {
+                if (game.cases[i][j] != 0) {
+                    var x = j * w;
+                    var y = i * h;
+
+                    if (loaded) {
+                        var imW = img.width / game.size;
+                        var imH = img.height / game.size;
+                        var sx = ((game.cases[i][j] - 1) % game.size) * imW;
+                        var sy = Math.floor((game.cases[i][j] - 1) / game.size) * imH;
+                        ctx.drawImage(img, sx, sy, imW, imH, x, y, w, h);
+
+                        if (show_help && !has_won) {
+                            ctx.fillStyle = "white";
+                            ctx.font = "bold 23px Arial";
+                            ctx.fillText(game.cases[i][j], x + 4, y + 20);
+                            ctx.fillStyle = "black";
+                            ctx.font = "bold 20px Arial";
+                            ctx.fillText(game.cases[i][j], x + 5, y + 19);
+                        }
+                        if (show_grid) {
+                            ctx.strokeStyle = "white";
+                            ctx.lineWidth = 2;
+                            ctx.strokeRect(x, y, w, h);
+                            ctx.strokeStyle = "black";
+                            ctx.lineWidth = 1;
+                            ctx.strokeRect(x, y, w, h);
+                        }
+                    } else {
+                        ctx.fillStyle = colors[game.cases[i][j] - 1];
+                        ctx.fillRect(x, y, w, h);
+                        ctx.fillStyle = "black";
+                        ctx.font = "bold 20px Arial";
+                        ctx.fillText(game.cases[i][j], x + w / 2 - 5, y + h / 2 + 5);
+                    }
+                }
+            }
+        }
+    }
+
+    function click(e) {
+        const rect = can.getBoundingClientRect();
+        var x = e.clientX - rect.left < 0 ? 0 : e.clientX - rect.left;
+        var y = e.clientY - rect.top < 0 ? 0 : e.clientY - rect.top;
```

Comment:
Before the fix, `x` and `y` could be negative, wich lead to `var i = -1` which is not correct